File: cli/src/huerise_cli/commands/alarms.py
from typing import Annotated
from uuid import UUID

import typer

from huerise_cli.client import api_client
from huerise_cli.config import Config
from huerise_cli.generated.api.alarms import (
    create_alarm as create_alarm_api,
    delete_alarm as delete_alarm_api,
    disable_alarm as disable_alarm_api,
    dismiss_alarm as dismiss_alarm_api,
    enable_alarm as enable_alarm_api,
    get_alarm as get_alarm_api,
    list_alarms as list_alarms_api,
    list_occurrences as list_occurrences_api,
    snooze_alarm as snooze_alarm_api,
)
from huerise_cli.generated.models import (
    AlarmCreate,
    AlarmRead,
    OccurrenceRead,
    ScheduleSchema,
    SnoozeRequest,
    Weekday,
)
from huerise_cli.generated.types import UNSET, Unset
from huerise_cli.output import (
    JsonOption,
    console,
    emit_json,
    emit_record,
    emit_table,
    unwrap,
)

app = typer.Typer(no_args_is_help=True, help="Manage sunrise alarms.")

_WEEKDAYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")


def _parse_days(days: list[str]) -> list[Weekday]:
    try:
        return [Weekday(_WEEKDAYS.index(day.lower())) for day in days]
    except ValueError as error:
        raise typer.BadParameter(
            f"{error}. Choose from: {', '.join(_WEEKDAYS)}"
        ) from error


def _format_schedule(schedule: ScheduleSchema) -> str:
    time = f"{schedule.hour:02d}:{schedule.minute:02d}"
    if isinstance(schedule.days, Unset) or not schedule.days:
        return f"{time} once"
    names = ",".join(_WEEKDAYS[int(day)] for day in schedule.days)
    return f"{time} {names}"


def _alarm_row(alarm: AlarmRead) -> dict[str, object]:
    return {
        "id": str(alarm.id),
        "label": alarm.label,
        "room": alarm.room_name,
        "schedule": _format_schedule(alarm.schedule),
        "enabled": alarm.is_enabled,
        "next": alarm.next_occurrence.isoformat() if alarm.next_occurrence else "-",
    }


def _render_alarm(alarm: AlarmRead, as_json: bool) -> None:
    if as_json:
        emit_json(alarm.to_dict())
    else:
        emit_record(_alarm_row(alarm))


def _occurrence_row(occurrence: OccurrenceRead) -> dict[str, object]:
    return {
        "id": str(occurrence.id),
        "scheduled_for": occurrence.scheduled_for.isoformat(),
        "state": occurrence.state,
        "snoozes": occurrence.snooze_count,
        "failure_reason": occurrence.failure_reason or "-",
    }


def _render_occurrence(occurrence: OccurrenceRead, as_json: bool) -> None:
    if as_json:
        emit_json(occurrence.to_dict())
    else:
        emit_record(_occurrence_row(occurrence))


@app.command("list")
def list_alarms(json_: JsonOption = False) -> None:
    """List every alarm."""
    with api_client(Config.from_env()) as client:
        alarms = unwrap(list_alarms_api.sync_detailed(client=client))

    if json_:
        emit_json([alarm.to_dict() for alarm in alarms])
    else:
        emit_table(
            [_alarm_row(alarm) for alarm in alarms], empty_message="No alarms yet."
        )


@app.command("create")
def create_alarm(
    label: Annotated[str, typer.Argument(help="Human-readable name for the alarm.")],
    room: Annotated[str, typer.Option(help="Room to run the sunrise scene in.")],
    hour: Annotated[int, typer.Option(min=0, max=23, help="Hour, 0-23.")],
    minute: Annotated[int, typer.Option(min=0, max=59, help="Minute, 0-59.")],
    timezone: Annotated[str, typer.Option(help="IANA timezone.")] = "Europe/Berlin",
    day: Annotated[
        list[str] | None,
        typer.Option(
            "--day",
            help=f"Weekday to repeat on ({', '.join(_WEEKDAYS)}). "
            "Omit for a one-off alarm.",
        ),
    ] = None,
    profile_id: Annotated[
        UUID | None,
        typer.Option(help="Alarm profile to use. Defaults to the default profile."),
    ] = None,
    json_: JsonOption = False,
) -> None:
    """Create a new alarm."""
    schedule = ScheduleSchema(
        hour=hour, minute=minute, timezone=timezone, days=_parse_days(day or [])
    )
    body = AlarmCreate(
        label=label,
        schedule=schedule,
        room_name=room,
        profile_id=profile_id if profile_id is not None else UNSET,
    )
    with api_client(Config.from_env()) as client:
        alarm = unwrap(create_alarm_api.sync_detailed(client=client, body=body))
    _render_alarm(alarm, json_)


@app.command("get")
def get_alarm(alarm_id: UUID, json_: JsonOption = False) -> None:
    """Show a single alarm."""
    with api_client(Config.from_env()) as client:
        alarm = unwrap(get_alarm_api.sync_detailed(client=client, alarm_id=alarm_id))
    _render_alarm(alarm, json_)


@app.command("enable")
def enable_alarm(alarm_id: UUID, json_: JsonOption = False) -> None:
    """Enable an alarm."""
    with api_client(Config.from_env()) as client:
        alarm = unwrap(enable_alarm_api.sync_detailed(client=client, alarm_id=alarm_id))
    _render_alarm(alarm, json_)


@app.command("disable")
def disable_alarm(alarm_id: UUID, json_: JsonOption = False) -> None:
    """Disable an alarm."""
    with api_client(Config.from_env()) as client:
        alarm = unwrap(
            disable_alarm_api.sync_detailed(client=client, alarm_id=alarm_id)
        )
    _render_alarm(alarm, json_)


@app.command("snooze")
def snooze_alarm(
    alarm_id: UUID,
    minutes: Annotated[int, typer.Option(min=1, max=60)] = 10,
    json_: JsonOption = False,
) -> None:
    """Snooze the alarm's current occurrence."""
    with api_client(Config.from_env()) as client:
        occurrence = unwrap(
            snooze_alarm_api.sync_detailed(
                client=client, alarm_id=alarm_id, body=SnoozeRequest(minutes=minutes)
            )
        )
    _render_occurrence(occurrence, json_)


@app.command("dismiss")
def dismiss_alarm(alarm_id: UUID, json_: JsonOption = False) -> None:
    """Dismiss the alarm's current occurrence."""
    with api_client(Config.from_env()) as client:
        occurrence = unwrap(
            dismiss_alarm_api.sync_detailed(client=client, alarm_id=alarm_id)
        )
    _render_occurrence(occurrence, json_)


@app.command("occurrences")
def list_occurrences(
    alarm_id: UUID,
    limit: Annotated[int, typer.Option(help="Max occurrences to return.")] = 20,
    json_: JsonOption = False,
) -> None:
    """List past and upcoming occurrences of an alarm."""
    with api_client(Config.from_env()) as client:
        occurrences = unwrap(
            list_occurrences_api.sync_detailed(
                client=client, alarm_id=alarm_id, limit=limit
            )
        )
    if json_:
        emit_json([occurrence.to_dict() for occurrence in occurrences])
    else:
        emit_table(
            [_occurrence_row(occurrence) for occurrence in occurrences],
            empty_message="No occurrences yet.",
        )


@app.command("delete")
def delete_alarm(
    alarm_id: UUID,
    yes: Annotated[
        bool, typer.Option("--yes", "-y", help="Skip confirmation.")
    ] = False,
) -> None:
    """Delete an alarm."""
    if not yes:
        typer.confirm(f"Delete alarm {alarm_id}?", abort=True)
    with api_client(Config.from_env()) as client:
        unwrap(delete_alarm_api.sync_detailed(client=client, alarm_id=alarm_id))
    console.print(f"[green]Deleted alarm {alarm_id}.[/green]")

File: cli/internal/commands/alarms.go
package commands

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/mathisarends/huerise/cli/internal/client"
)

var weekdayNames = []string{"mon", "tue", "wed", "thu", "fri", "sat", "sun"}

type alarmsCommand struct {
	List        alarmsListCommand        `cmd:"" help:"List every alarm."`
	Create      alarmsCreateCommand      `cmd:"" help:"Create a new alarm."`
	Get         alarmsGetCommand         `cmd:"" help:"Show a single alarm."`
	Enable      alarmsEnableCommand      `cmd:"" help:"Enable an alarm."`
	Disable     alarmsDisableCommand     `cmd:"" help:"Disable an alarm."`
	Snooze      alarmsSnoozeCommand      `cmd:"" help:"Snooze the current occurrence."`
	Dismiss     alarmsDismissCommand     `cmd:"" help:"Dismiss the current occurrence."`
	Occurrences alarmsOccurrencesCommand `cmd:"" help:"List recent occurrences."`
	Delete      alarmsDeleteCommand      `cmd:"" help:"Delete an alarm."`
}

type alarmsListCommand struct{}

type alarmsCreateCommand struct {
	Label     string   `arg:"" help:"Human-readable alarm name."`
	Room      string   `required:"" help:"Room to run the sunrise scene in."`
	Hour      int      `required:"" help:"Hour (0-23)."`
	Minute    int      `required:"" help:"Minute (0-59)."`
	Timezone  string   `default:"Europe/Berlin" help:"IANA timezone."`
	Day       []string `help:"Weekday to repeat on; repeat for multiple days (mon-sun)."`
	ProfileID string   `name:"profile-id" help:"Alarm profile UUID; defaults to the default profile."`
}

type alarmsGetCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
}

type alarmsEnableCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
}

type alarmsDisableCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
}

type alarmsSnoozeCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
	Minutes int       `default:"10" help:"Snooze duration (1-60 minutes)."`
}

type alarmsDismissCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
}

type alarmsOccurrencesCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
	Limit   int       `default:"20" help:"Maximum occurrences to return."`
}

type alarmsDeleteCommand struct {
	AlarmID uuid.UUID `arg:"" name:"alarm-id"`
	Yes     bool      `short:"y" help:"Skip confirmation."`
}

func (alarmsListCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	alarms, err := apiClient.ListAlarms(runtime.ctx)
	if err != nil {
		return err
	}
	return runtime.output(alarms, func() error {
		rows := make([][]string, 0, len(alarms))
		for _, alarm := range alarms {
			rows = append(rows, alarmRow(alarm))
		}
		return writeTable(runtime.stdout, []string{"ID", "LABEL", "ROOM", "SCHEDULE", "ENABLED", "NEXT"}, rows, "No alarms yet.")
	})
}

func (command alarmsCreateCommand) Run(runtime *Runtime) error {
	days, err := parseWeekdays(command.Day)
	if err != nil {
		return err
	}
	if command.Hour < 0 || command.Hour > 23 {
		return &commandError{Code: "usage", Message: "hour must be between 0 and 23", ExitCode: 2}
	}
	if command.Minute < 0 || command.Minute > 59 {
		return &commandError{Code: "usage", Message: "minute must be between 0 and 59", ExitCode: 2}
	}
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	room, err := resolveRoom(runtime, apiClient, command.Room)
	if err != nil {
		return err
	}
	request := &client.AlarmCreate{
		Label: command.Label,
		Schedule: client.ScheduleSchema{
			Hour: command.Hour, Minute: command.Minute,
			Timezone: client.NewOptString(command.Timezone), Days: days,
		},
		RoomID:   room.ID,
		RoomName: room.Name,
	}
	if command.ProfileID != "" {
		profileID, parseErr := uuid.Parse(command.ProfileID)
		if parseErr != nil {
			return &commandError{Code: "usage", Message: "invalid --profile-id: " + parseErr.Error(), ExitCode: 2}
		}
		request.ProfileID = client.NewOptNilUUID(profileID)
	}
	response, err := apiClient.CreateAlarm(runtime.ctx, request)
	if err != nil {
		return err
	}
	return handleAlarmResponse(runtime, "create alarm", response)
}

func (command alarmsGetCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.GetAlarm(runtime.ctx, client.GetAlarmParams{AlarmID: command.AlarmID})
	if err != nil {
		return err
	}
	return handleAlarmResponse(runtime, "get alarm", response)
}

func (command alarmsEnableCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.EnableAlarm(runtime.ctx, client.EnableAlarmParams{AlarmID: command.AlarmID})
	if err != nil {
		return err
	}
	return handleAlarmResponse(runtime, "enable alarm", response)
}

func (command alarmsDisableCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.DisableAlarm(runtime.ctx, client.DisableAlarmParams{AlarmID: command.AlarmID})
	if err != nil {
		return err
	}
	return handleAlarmResponse(runtime, "disable alarm", response)
}

func (command alarmsSnoozeCommand) Run(runtime *Runtime) error {
	if command.Minutes < 1 || command.Minutes > 60 {
		return &commandError{Code: "usage", Message: "minutes must be between 1 and 60", ExitCode: 2}
	}
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.SnoozeAlarm(runtime.ctx,
		&client.SnoozeRequest{Minutes: client.NewOptInt(command.Minutes)},
		client.SnoozeAlarmParams{AlarmID: command.AlarmID},
	)
	if err != nil {
		return err
	}
	return handleOccurrenceResponse(runtime, "snooze alarm", response)
}

func (command alarmsDismissCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.DismissAlarm(runtime.ctx, client.DismissAlarmParams{AlarmID: command.AlarmID})
	if err != nil {
		return err
	}
	return handleOccurrenceResponse(runtime, "dismiss alarm", response)
}

func (command alarmsOccurrencesCommand) Run(runtime *Runtime) error {
	if command.Limit < 1 {
		return &commandError{Code: "usage", Message: "limit must be at least 1", ExitCode: 2}
	}
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.ListOccurrences(runtime.ctx, client.ListOccurrencesParams{
		AlarmID: command.AlarmID, Limit: client.NewOptInt(command.Limit),
	})
	if err != nil {
		return err
	}
	switch result := response.(type) {
	case *client.ListOccurrencesOKApplicationJSON:
		occurrences := []client.OccurrenceRead(*result)
		return runtime.output(occurrences, func() error {
			rows := make([][]string, 0, len(occurrences))
			for _, occurrence := range occurrences {
				rows = append(rows, occurrenceRow(occurrence))
			}
			return writeTable(runtime.stdout, []string{"ID", "SCHEDULED FOR", "STATE", "SNOOZES", "FAILURE"}, rows, "No occurrences yet.")
		})
	case *client.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected list occurrences response %T", response)
	}
}

func (command alarmsDeleteCommand) Run(runtime *Runtime) error {
	if !command.Yes {
		if err := confirmDelete(runtime, "alarm", command.AlarmID); err != nil {
			return err
		}
	}
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := apiClient.DeleteAlarm(runtime.ctx, client.DeleteAlarmParams{AlarmID: command.AlarmID})
	if err != nil {
		return err
	}
	switch result := response.(type) {
	case *client.DeleteAlarmNoContent:
		return runtime.output(map[string]any{"deleted": true, "id": command.AlarmID.String()}, func() error {
			_, err := fmt.Fprintf(runtime.stdout, "Deleted alarm %s.\n", command.AlarmID)
			return err
		})
	case *client.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected delete alarm response %T", response)
	}
}

func parseWeekdays(values []string) ([]client.Weekday, error) {
	days := make([]client.Weekday, 0, len(values))
	seen := map[client.Weekday]bool{}
	for _, value := range values {
		index := -1
		for candidate, name := range weekdayNames {
			if strings.EqualFold(value, name) {
				index = candidate
				break
			}
		}
		if index < 0 {
			return nil, &commandError{Code: "usage", Message: fmt.Sprintf("invalid weekday %q", value), Hint: "Choose from: " + strings.Join(weekdayNames, ", "), ExitCode: 2}
		}
		day := client.Weekday(index)
		if !seen[day] {
			days = append(days, day)
			seen[day] = true
		}
	}
	return days, nil
}

func handleAlarmResponse(runtime *Runtime, operation string, response any) error {
	switch result := response.(type) {
	case *client.AlarmRead:
		return writeAlarm(runtime, result)
	case *client.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected %s response %T", operation, response)
	}
}

func handleOccurrenceResponse(runtime *Runtime, operation string, response any) error {
	switch result := response.(type) {
	case *client.OccurrenceRead:
		return writeOccurrence(runtime, result)
	case *client.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected %s response %T", operation, response)
	}
}

func writeAlarm(runtime *Runtime, alarm *client.AlarmRead) error {
	return runtime.output(alarm, func() error {
		row := alarmRow(*alarm)
		return writeRecord(runtime.stdout,
			recordField{Name: "id", Value: row[0]},
			recordField{Name: "label", Value: row[1]},
			recordField{Name: "room", Value: row[2]},
			recordField{Name: "schedule", Value: row[3]},
			recordField{Name: "enabled", Value: row[4]},
			recordField{Name: "next", Value: row[5]},
		)
	})
}

func alarmRow(alarm client.AlarmRead) []string {
	next := "-"
	if value, ok := alarm.NextOccurrence.Get(); ok {
		next = value.Format(time.RFC3339)
	}
	return []string{
		alarm.ID.String(), alarm.Label, alarm.RoomName, formatSchedule(alarm.Schedule),
		strconv.FormatBool(alarm.IsEnabled), next,
	}
}

func formatSchedule(schedule client.ScheduleSchema) string {
	value := fmt.Sprintf("%02d:%02d", schedule.Hour, schedule.Minute)
	if len(schedule.Days) == 0 {
		return value + " once"
	}
	days := make([]string, 0, len(schedule.Days))
	for _, day := range schedule.Days {
		index := int(day)
		if index >= 0 && index < len(weekdayNames) {
			days = append(days, weekdayNames[index])
		}
	}
	return value + " " + strings.Join(days, ",")
}

func writeOccurrence(runtime *Runtime, occurrence *client.OccurrenceRead) error {
	return runtime.output(occurrence, func() error {
		row := occurrenceRow(*occurrence)
		return writeRecord(runtime.stdout,
			recordField{Name: "id", Value: row[0]},
			recordField{Name: "scheduled_for", Value: row[1]},
			recordField{Name: "state", Value: row[2]},
			recordField{Name: "snoozes", Value: row[3]},
			recordField{Name: "failure", Value: row[4]},
		)
	})
}

func occurrenceRow(occurrence client.OccurrenceRead) []string {
	return []string{
		occurrence.ID.String(), occurrence.ScheduledFor.Format(time.RFC3339), string(occurrence.State),
		strconv.Itoa(occurrence.SnoozeCount), occurrence.FailureReason.Or("-"),
	}
}

func confirmDelete(runtime *Runtime, resource string, id uuid.UUID) error {
	if runtime.root.NoInput || !isTerminal(runtime.stdin) {
		return &commandError{Code: "input_required", Message: "deleting a " + resource + " requires confirmation", Hint: "Pass --yes to skip confirmation.", ExitCode: 2}
	}
	_, _ = fmt.Fprintf(runtime.stderr, "Delete %s %s? [y/N] ", resource, id)
	answer, err := bufio.NewReader(runtime.stdin).ReadString('\n')
	if err != nil {
		return fmt.Errorf("read confirmation: %w", err)
	}
	answer = strings.TrimSpace(strings.ToLower(answer))
	if answer != "y" && answer != "yes" {
		return &commandError{Code: "cancelled", Message: "delete cancelled", ExitCode: 1}
	}
	return nil
}

func isTerminal(reader any) bool {
	file, ok := reader.(*os.File)
	if !ok {
		return false
	}
	info, err := file.Stat()
	return err == nil && info.Mode()&os.ModeCharDevice != 0
}
